File: tools/factory-data-architect/src/types/recipe.ts
// Recipe Editor Types

// Ingredient type (Item or Tag)
export type IngredientType =
  | { type: "Item"; value: string }
  | { type: "Tag"; value: string };

// Ingredient
export interface Ingredient {
  ingredient_type: IngredientType;
  amount: number;
}

// Product type (Item or Fluid)
export type ProductType =
  | { type: "Item"; value: string }
  | { type: "Fluid"; value: string };

// Product (result)
export interface Product {
  product_type: ProductType;
  amount: number;
  chance: number; // 0.0 - 1.0
}

// Machine types
export type MachineType =
  | "Assembler"
  | "Mixer"
  | "Press"
  | "Furnace"
  | "Crusher"
  | "Centrifuge"
  | "ChemicalReactor"
  | "Packager"
  | { Custom: string };

// Recipe definition (matches Rust RecipeDef)
export interface RecipeDef {
  id: string;
  machine_type: MachineType;
  ingredients: Ingredient[];
  results: Product[];
  process_time: number;
  stress_impact: number;
  i18n_key: string;
}

// Catalog entry for palette
export interface CatalogEntry {
  id: string;
  name: string;
  icon_path: string | null;
}

// Machine catalog entry
export interface MachineCatalogEntry {
  id: string;
  name: string;
  machine_type: MachineType;
  input_slots: number;
  output_slots: number;
}

// Asset catalog (for palette) - matches Rust backend
export interface AssetCatalog {
  items: CatalogEntry[];
  fluids: CatalogEntry[];
  machines: MachineCatalogEntry[];
  tags: string[];
}

// ============ React Flow Node Types ============

// Base node data
export interface BaseNodeData {
  label: string;
}

// Source Node (材料)
export interface SourceNodeData extends BaseNodeData {
  nodeType: "source";
  itemType: "item" | "fluid" | "tag";
  itemId: string;
  amount: number;
}

// Machine Node (加工機械)
export interface MachineNodeData extends BaseNodeData {
  nodeType: "machine";
  machineType: MachineType;
  machineId?: string;
  inputSlots: number;
  outputSlots: number;
}

// Result Node (成果物)
export interface ResultNodeData extends BaseNodeData {
  nodeType: "result";
  itemType: "item" | "fluid";
  itemId: string;
  amount: number;
  chance: number;
}

// Union type for all node data
export type RecipeNodeData = SourceNodeData | MachineNodeData | ResultNodeData;

// Node types for React Flow
export const NODE_TYPES = {
  source: "source",
  machine: "machine",
  result: "result",
} as const;

// Palette item for drag & drop
export interface PaletteItem {
  type: "source" | "machine" | "result";
  itemType?: "item" | "fluid" | "tag";
  id?: string;
  machineType?: MachineType;
}

// Helper functions
export function createDefaultRecipe(id: string): RecipeDef {
  return {
    id,
    machine_type: "Assembler",
    ingredients: [],
    results: [],
    process_time: 1.0,
    stress_impact: 0.0,
    i18n_key: `recipe.${id}`,
  };
}

export function createSourceNodeData(
  itemType: "item" | "fluid" | "tag",
  itemId: string,
  name: string
): SourceNodeData {
  return {
    nodeType: "source",
    label: name,
    itemType,
    itemId,
    amount: 1,
  };
}

export function createMachineNodeData(
  machineType: MachineType,
  machineId: string,
  name: string,
  inputSlots: number,
  outputSlots: number
): MachineNodeData {
  return {
    nodeType: "machine",
    label: name,
    machineType,
    machineId,
    inputSlots,
    outputSlots,
  };
}

export function createResultNodeData(
  itemType: "item" | "fluid",
  itemId: string,
  name: string
): ResultNodeData {
  return {
    nodeType: "result",
    label: name,
    itemType,
    itemId,
    amount: 1,
    chance: 1.0,
  };
}

File: tools/factory-data-architect/src-tauri/src/recipe.rs
use serde::{Deserialize, Serialize};

/// 材料の種類（アイテムまたはタグ）
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "value")]
pub enum IngredientType {
    /// 特定のアイテムID
    Item(String),
    /// 鉱石辞書タグ（複数アイテムを許容）
    Tag(String),
}

/// レシピの材料
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Ingredient {
    /// アイテムまたはタグ
    pub ingredient_type: IngredientType,
    /// 必要数量
    pub amount: u32,
}

/// 生成物の種類
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "value")]
pub enum ProductType {
    /// アイテム
    Item(String),
    /// 流体
    Fluid(String),
}

/// レシピの生成物
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Product {
    /// アイテムまたは流体
    pub product_type: ProductType,
    /// 生成数量
    pub amount: u32,
    /// 生成確率 (0.0 - 1.0)
    #[serde(default = "default_chance")]
    pub chance: f32,
}

fn default_chance() -> f32 {
    1.0
}

/// 機械タイプ
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum MachineType {
    Assembler,
    Mixer,
    Press,
    Furnace,
    Crusher,
    Centrifuge,
    ChemicalReactor,
    Packager,
    Custom(String),
}

impl Default for MachineType {
    fn default() -> Self {
        MachineType::Assembler
    }
}

/// レシピ定義
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RecipeDef {
    /// レシピID
    pub id: String,
    /// 使用する機械タイプ
    pub machine_type: MachineType,
    /// 材料リスト
    pub ingredients: Vec<Ingredient>,
    /// 生成物リスト
    pub results: Vec<Product>,
    /// 加工時間（秒）
    pub process_time: f32,
    /// 消費応力
    #[serde(default)]
    pub stress_impact: f32,
    /// 説明用i18nキー
    #[serde(default)]
    pub i18n_key: String,
}

impl RecipeDef {
    pub fn new(id: String) -> Self {
        Self {
            i18n_key: format!("recipe.{}", id),
            id,
            machine_type: MachineType::default(),
            ingredients: Vec::new(),
            results: Vec::new(),
            process_time: 1.0,
            stress_impact: 0.0,
        }
    }
}

/// アセットツリーのノード
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum AssetNode {
    /// ディレクトリ
    Directory {
        name: String,
        children: Vec<AssetNode>,
    },
    /// アイテム定義
    Item {
        id: String,
        name: String,
        path: String,
    },
    /// 流体定義
    Fluid {
        id: String,
        name: String,
    },
    /// 機械定義
    Machine {
        id: String,
        name: String,
        machine_type: MachineType,
    },
}

/// パレット用のアセットカタログ
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct AssetCatalog {
    pub items: Vec<CatalogEntry>,
    pub fluids: Vec<CatalogEntry>,
    pub machines: Vec<MachineCatalogEntry>,
    pub tags: Vec<String>,
}

/// カタログエントリ
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CatalogEntry {
    pub id: String,
    pub name: String,
    pub icon_path: Option<String>,
}

/// 機械カタログエントリ
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MachineCatalogEntry {
    pub id: String,
    pub name: String,
    pub machine_type: MachineType,
    pub input_slots: u32,
    pub output_slots: u32,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_recipe_serialization() {
        let recipe = RecipeDef {
            id: "iron_plate".to_string(),
            machine_type: MachineType::Press,
            ingredients: vec![Ingredient {
                ingredient_type: IngredientType::Item("iron_ingot".to_string()),
                amount: 1,
            }],
            results: vec![Product {
                product_type: ProductType::Item("iron_plate".to_string()),
                amount: 1,
                chance: 1.0,
            }],
            process_time: 2.0,
            stress_impact: 4.0,
            i18n_key: "recipe.iron_plate".to_string(),
        };

        let json = serde_json::to_string_pretty(&recipe).unwrap();
        let deserialized: RecipeDef = serde_json::from_str(&json).unwrap();
        assert_eq!(recipe.id, deserialized.id);
    }
}
